### DISCUSSION TITLE
Add a way (command or step) to early-exit the job and set check conclusion

### DISCUSSION THREAD

**Discussion by Jackenmen** [community/community#82744] [Actions]
### Select Topic Area

Product Feedback

### Body

Currently, to skip the rest of the steps during a job, an `if` condition has to be added to all of the steps or steps have to be moved to another job.

It'd be nice to have [a command](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions) or a step to early-exit (premature, prematurely, graceful, gracefully) (end, finish, terminate, termination, stop, halt, skip, cancel, discontinue, fail) the job during a step without failing the job:

```
::exit::
```

It'd also be nice to set the conclusion of the early-exited job:

```
::exit::failure
```

(equivalent to `exit 1`) or

```
::exit::neutral
```
 
(equivalent to `exit 78` [in Actions v1](https://github.community/t/github-actions-neutral-exit-code-is-incorrectly-interpreted-as-failure/16088)) and [other check conclusions](https://developer.github.com/v3/checks/runs/#parameters).

This would help save build time for both GitHub and customers.

### Related StackOverflow and GitHub Community posts

* [How to fail a job in Github Actions?](https://stackoverflow.com/q/57903836)
* [How to force to exit in Github Actions step](https://stackoverflow.com/q/60589373)
* [How to write a GitHub Action with a neutral Check Run status](https://stackoverflow.com/q/57687369)
* [How to perform an early clean exit of a job ?](https://github.community/t/how-to-perform-an-early-clean-exit-of-a-job/17531)
* [Early successful termination of job](https://github.community/t/early-successful-termination-of-job/16500)
* [Github Action Skip Steps/Fall Through Syntax](https://stackoverflow.com/q/60382188)

### Related issues

* [Has the `setNeutral` method been removed? toolkit#146](https://github.com/actions/toolkit/issues/146)
* [veggiemonk/skip-commit](https://github.com/veggiemonk/skip-commit)

### Same functionality in other CIs

* CircleCI: [`circleci-agent step halt`](https://circleci.com/docs/2.0/configuration-reference/#ending-a-job-from-within-a-step)
* Travis CI: [`travis_terminate`](https://github.com/travis-ci/travis-build/blob/master/lib/travis/build/bash/travis_terminate.bash?rgh-link-date=2020-08-16T10%3A40%3A43Z)
* AppVeyor: [`appveyor exit`](https://www.appveyor.com/docs/build-configuration/#build-pipeline) ([Linux](https://www.appveyor.com/docs/getting-started-with-appveyor-for-linux/#premature-termination-of-the-build-process))
* Jenkins: [`currentBuild.result`](https://devops.stackexchange.com/q/885)
* GitLab CI: [`exit 0`](https://gitlab.com/gitlab-org/gitlab-foss/-/issues/18041)

---

_Originally posted by @ylemkimon in https://github.com/actions/runner/issues/662#issue-679731420_

This issue has received A LOT of upvotes (800+ at the time of writing) on the actions/runner repository but has since been closed with a request to redirect this feedback to this forum which is what this newly made discussion is meant to achieve:
https://github.com/actions/runner/issues/662

**Comment by jpbochi** [comment]
Currently, https://github.com/actions/toolkit/tree/main/packages/core#exit-codes says:

> Note that `setNeutral` is not yet implemented in actions V2 but equivalent functionality is being planned.

The docs at https://docs.github.com/en/actions/creating-actions/setting-exit-codes-for-actions only mentions `success` and `failure`.

---

One possible workaround would be to update the check run after it fails (See https://docs.github.com/en/rest/checks/runs#update-a-check-run). Unfortunately, that can't be done with a regular user token, only with [GitHub Apps](https://docs.github.com/en/developers/apps).

**Comment by jazanne** [comment]
This is something that has come for me a lot when working on workflows, would REALLY love to see it implemented

**Comment by punkbit** [comment]
If you're looking for a solution, at least to cancel:

```
gh run cancel ${{ github.run_id }}
exit 1
```

Results in:

```
✓ Request to cancel workflow XXXXXXX submitted.
Error: Process completed with exit code 1.
```

⚠️ The problem is that if you have concurrent builds, it'll cancel the whole job.

**Reply by Ferroin** [reply]
The second problem with this is that it only works if you don’t care about the result status for the job.

**Reply by elibarzilay** [reply]
The third problem is that it's a ****ing horrible way to cancel a build, as seen by the usual obligatory `sleep` that comes with it to make sure you don't get to do more stuff after the cancellation.

**Comment by vt89** [comment]
It is a must to avoid repetitive if conditions on all the following steps and improve readability!

**Comment by Steve-Glass** [comment]
👋 Chiming in from the Actions product team. I may roll this request up into a general backlog item for conditional operators. I think [this is a similar ask](https://github.com/orgs/community/discussions/75928). This is something that may be considered for this year. I will report back in October after planning to inform this group of the outcome.

**Reply by rmacklin** [reply]
Can you please elaborate? I'm not sure I follow how that ask would address this. Specifically, not sure how it would address the need to say "do not run any subsequent steps" without having to add an `if:` to each subsequent step's yaml.

**Reply by jazanne** [reply]
same question - the issues you linked doesn't solve the issue I'm facing

**Reply by Steve-Glass** [reply]
@rmacklin My apologies, this was meant for another thread. Symptom of having many tabs open at once 😞 I plan on taking a look at this request in the next week  or so, just focusing on other areas at the moment.

**Reply by rmacklin** [reply]
Gotcha, thanks for clarifying! Stoked to hear you'll be looking at this issue! :)

**Reply by owattley-rotageek** [reply]
Any update @Steve-Glass ?

**Reply by ubuntudroid** [reply]
It's been a half a year since your last comment, @Steve-Glass. Did this end up on the near term roadmap?

**Reply by pmaAXA** [reply]
Look taken @Steve-Glass?

**Comment by dwilches** [comment]
Has this feature been released? I can't find a way to avoid the "if" in each of the steps of a GHA to make the job exit as soon as a step determined there is nothing to do. Is there a planned release for this feature?

(Maybe @Steve-Glass)

**Reply by braboobssiere** [reply]
you can early exit with exit 1 but it will show as failure on status (if you don't care to differentiate between actual failure and failure from early exit then it's fine)

**Reply by meins111** [reply]
Also still waiting on this. 
Current use-case:
Have a build-deploy workflow with manual dispatch. User input defines which component(s) are supposed to be build and deployed.
Implemented as a matrix action, and a first step to check whether the matrix combination is "selected" based on user input.
Doing an exit 1 will show (partial) failed result in the deploy action - no one wants that...

**Reply by Ferroin** [reply]
@meins111 For that specific use-case, had you considered dynamically generating the build matrix instead of skipping the parts you don’t need?

Something like:

```yaml
jobs:
  matrix:
    name: Prepare build matrix
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.set-matrix.outputs.matrix }}
    steps:
      - uses: actions/checkout@v4
      - id: set-matrix
        run: |
          matrix="$(./.github/scripts/generate-matrix.sh)"
          echo "${matrix}"
          echo "matrix=${matrix}" >> "${GITHUB_OUTPUT}"

  build:
    name: Build
    needs:
      - matrix
    strategy:
      matrix: ${{ fromJson(needs.matrix.outputs.matrix) }}
```

And just have `.github/scripts/generate-matrix.sh` produce a JSON representation of the build matrix on stdout when run.

Speaking from experience, this works very well provided you don’t need to make the jobs generated from the matrix required checks on a PR.

**Reply by pcharlan** [reply]
> Speaking from experience, [building the matrix dynamically] works very well provided you don’t need to make the jobs generated from the matrix required checks on a PR.

You can make a do-nothing job named something like "Tests Complete", make it dependent on the matrix job, and add that job as a required PR check.  It will pass if and only if all the Matrix jobs passed.

This may not work for every use case, but we use it to solve this problem and it works well.
